src/mlpack/core/tree/cellbound_impl.hpp

```diff
@@ -0,0 +1,970 @@
+  }
+  else
+    return maxSum;
+}
+
+/**
+ * Computes maximum distance.
+ */
+template<typename MetricType, typename ElemType>
+inline ElemType CellBound<MetricType, ElemType>::MaxDistance(
+    const CellBound& other)
+    const
+{
+  ElemType maxSum = std::numeric_limits<ElemType>::lowest();
+
+  Log::Assert(dim == other.dim);
+
+  ElemType v;
+  for (size_t i = 0; i < numBounds; i++)
+    for (size_t j = 0; j < other.numBounds; j++)
+    {
+      ElemType sum = 0;
+      for (size_t d = 0; d < dim; d++)
+      {
+        v = std::max(fabs(other.hiBound(d, j) - loBound(d, i)),
+            fabs(hiBound(d, i) - other.loBound(d, j)));
+
+        // The compiler should optimize out this if statement entirely.
+        if (MetricType::Power == 1)
+          sum += v; // v is non-negative.
+        else if (MetricType::Power == 2)
+          sum += v * v;
+        else
+          sum += std::pow(v, (ElemType) MetricType::Power);
+      }
+
+      if (sum > maxSum)
+        maxSum = sum;
+    }
+
+  // The compiler should optimize out this if statement entirely.
+  if (MetricType::TakeRoot)
+  {
+    if (MetricType::Power == 1)
+      return maxSum;
+    else if (MetricType::Power == 2)
+      return (ElemType) std::sqrt(maxSum);
+    else
+      return (ElemType) pow((double) maxSum, 1.0 / (double) MetricType::Power);
+  }
+  else
+    return maxSum;
+}
+
+/**
+ * Calculates minimum and maximum bound-to-bound squared distance.
+ */
+template<typename MetricType, typename ElemType>
+inline math::RangeType<ElemType>
+CellBound<MetricType, ElemType>::RangeDistance(
+    const CellBound& other) const
+{
+  ElemType minLoSum = std::numeric_limits<ElemType>::max();
+  ElemType maxHiSum = std::numeric_limits<ElemType>::lowest();
+
+  Log::Assert(dim == other.dim);
+
+  ElemType v1, v2, vLo, vHi;
+
+  for (size_t i = 0; i < numBounds; i++)
+    for (size_t j = 0; j < other.numBounds; j++)
+    {
+      ElemType loSum = 0;
+      ElemType hiSum = 0;
+      for (size_t d = 0; d < dim; d++)
+      {
+        v1 = other.loBound(d, j) - hiBound(d, i);
+        v2 = loBound(d, i) - other.hiBound(d, j);
+        // One of v1 or v2 is negative.
+        if (v1 >= v2)
+        {
+          vHi = -v2; // Make it nonnegative.
+          vLo = (v1 > 0) ? v1 : 0; // Force to be 0 if negative.
+        }
+        else
+        {
+          vHi = -v1; // Make it nonnegative.
+          vLo = (v2 > 0) ? v2 : 0; // Force to be 0 if negative.
+        }
+
+        // The compiler should optimize out this if statement entirely.
+        if (MetricType::Power == 1)
+        {
+          loSum += vLo; // vLo is non-negative.
+          hiSum += vHi; // vHi is non-negative.
+        }
+        else if (MetricType::Power == 2)
+        {
+          loSum += vLo * vLo;
+          hiSum += vHi * vHi;
+        }
+        else
+        {
+          loSum += std::pow(vLo, (ElemType) MetricType::Power);
+          hiSum += std::pow(vHi, (ElemType) MetricType::Power);
+        }
```

Comment:
You could probably do early termination here too but the condition would be a but stricter.